### DISCUSSION TITLE
Preventing failing migrations in CICD?

### DISCUSSION THREAD

**Discussion by m1212e** [prisma/prisma#19843] [Q&A] [answered]
Hi, I'm new to prisma and was wondering how other folks prevent the following from happening:

I create a table with a few columns. I create the migration, commit and deploy via my CICD pipeline. The prod db gets migrated according to my new pushes.

Now I add a field which cannot be null and also has no default value. I commit and push again, this time the migration fails because there is already data which does not have the new field and since there is no default value it cannot be set automatically either.

I add another commit with the default value added, but since the migrations are done sequentially, it will still fail.

I remove the broken migration and the post added default value migration and create a new one, which combines the new field and default value.

But now I have to manually access the db and revoke the failed migration.

How do I prevent this from happening in the first place?

**Comment by nurul3101** [comment]
Hello @m1212e 👋 

If you are adding a new field that cannot be null, it is recommended to first allow it to be null or set a default value as part of the initial migration. This ensures that existing rows in the database will not cause the migration to fail.

You can then create a data migration script that will populate the new field for existing rows. This can be a separate script that's executed as part of the migration.

After you are sure that all existing rows have been populated with values in the new field, you can create another migration to alter the column to be non-nullable.

To prevent this from happening in the first place, you can enforce that when adding a non-nullable column, you combine the steps of adding the column with a default value or an intermediate nullable state, populating the existing rows, and setting it to non-nullable if required, all within a single migration script. This helps to ensure the migration is atomic and either completely succeeds or fails without partial changes.

Hope this helps! Please let me know in case you have any queries, if not please feel free to  **mark this Discussion as answered** to indicate that it has been resolved.

**Reply by m1212e** [reply]
Thank you that was very helpful. "you can enforce that when adding" did you mean actually enforcing via some kind of rule/config etc. or by telling people to do so?
Ideally a mindless/uncareful dev should just not be able to push something like I mentioned. Is that possible?

**Reply by nurul3101** [reply, marked as answer]
You can use tools like [Husky](https://typicode.github.io/husky/) to enforce checks before commits or pushes. You can write a script that scans the migration files for patterns such as adding non-nullable columns. The script can check if the same file also contains default values or makes the column nullable initially. If not, the script can block the commit or push.

**Reply by m1212e** [reply]
Ok, thank you. Thought about that aswell but seemed a bit too hacky for me. But I'll look into that, thanks!
